crates/bevy_asset/src/server/mod.rs

```diff
@@ -99,16 +99,35 @@ impl AssetServer {
     /// Registers a new [`AssetLoader`]. [`AssetLoader`]s must be registered before they can be used.
     pub fn register_loader<L: AssetLoader>(&self, loader: L) {
         let mut loaders = self.data.loaders.write();
-        let loader_index = loaders.values.len();
-        loaders
-            .type_name_to_index
-            .insert(std::any::type_name::<L>(), loader_index);
+        let type_name = std::any::type_name::<L>();
+        let loader = Arc::new(loader);
+        let (loader_index, is_new) =
+            if let Some(index) = loaders.preregistered_loaders.remove(type_name) {
+                (index, false)
+            } else {
+                (loaders.values.len(), true)
+            };
         for extension in loader.extensions() {
             loaders
                 .extension_to_index
                 .insert(extension.to_string(), loader_index);
         }
-        loaders.values.push(Arc::new(loader));
+
+        if is_new {
+            loaders.type_name_to_index.insert(type_name, loader_index);
+            loaders.values.push(MaybeAssetLoader::Ready(loader));
+        } else {
+            let maybe_loader = std::mem::replace(
+                &mut loaders.values[loader_index],
+                MaybeAssetLoader::Ready(loader.clone()),
+            );
+            match maybe_loader {
+                MaybeAssetLoader::Ready(_) => unreachable!(),
+                MaybeAssetLoader::Pending { sender, .. } => {
+                    let _ = sender.send_blocking(loader);
```

Comment:
mentioned on discord as well, but just to ensure it doesn't get missed: sending to async_channel results in only a single receiver getting unblocked / receiving the result, so only one waiting asset will get loaded with this approach. 

in #9453 we amended it to use `sender.close()` which worked fine there as we didn't use the actual result. here, i guess you need to use something like an async mutex, or a `tokio::sync::broadcast` channel, or you could stick with `sender.close()` and make the waiting tasks reacquire the loader after getting unblocked.

Reply:
Thanks for posting here too! I'm already using async_broadcast for this sort of thing in the asset processor. Should work well for this use case.

Reply:
Just made the swap. Only major change is `send_blocking` is replaced by calling `sender.await` in a detached task on the IoTaskPool